src/main/java/org/dependencytrack/util/NotificationUtil.java

```diff
@@ -108,6 +110,29 @@ public static void analyzeNotificationCriteria(QueryManager qm, Vulnerability vu
         }
     }
 
+    public static void analyzeNotificationCriteria(QueryManager qm, Vulnerability vulnerability, VulnerabilityUpdateDiff vulnerabilityUpdateDiff) {
+        // Vulnerabilities are not guaranteed to include component relationships depending on where the event was generated
+        final Vulnerability completeVulnerability = qm.getVulnerabilityByVulnId(vulnerability.getSource(), vulnerability.getVulnId());
+        final List<Component> components = completeVulnerability.getComponents();
```

Comment:
Fetching all associated components at once can be very costly in large portfolios, especially since only one of them is actually used. Do we need to include this info (see comment below as well)?

Reply:
This can absolutely be optimised if we don't need component information in the notification. It think it would be more efficient to include the affected projects instead, and use `qm.getAffectedProjects(vuln)` much like the REST API does in `VulnerabilityResource.getAffectedProject()` as we can then reuse this for filtering in `NotificationRouter`.

Reply:
`qm.getAffectedProjects(vuln)` actually calls `vuln.getComponents()` in the query manager anyway (which makes sense now that I've thought about it properly). Even if we don't include affected projects or components in the notification subject, we still need the affected project list to be able to limit notifications by project (or tags in the future). On the other hand, if we don't filter by affected projects then we might emit a large number of irrelevant notifications whenever there is a vulnerability database update, which also has performance implications.